### PR TITLE
Make NPC dialogue an overlay

### PR DESCRIPTION
**Pull request by magicus**
Also fixes bug in ChatManager and FontRenderer.

### CODE REVIEW

File: common/src/main/java/com/wynntils/mc/render/FontRenderer.java
Comment (magicus) [inline common/src/main/java/com/wynntils/mc/render/FontRenderer.java:163]: You know what, Qodana? I'm starting to get really pissed off with you. There is no `StringBuild.prepend()`. This rule of yours should have checked the order of the arguments to the `+`, not just naively assume that we don't know how to use a StringBuilder.

**Review by kristofbolyai — COMMENTED** [review]
This looks good. One thing I would like to confirm before I approve this though. At the moment, if we disable this feature (and the debug one), do we get the old, unmodified, Wynncraft style NPC messages (with no modification, including the "shift to continue", and the grayed out stuff)? 

If yes, this looks good!. If not, maybe we should add it in this PR. 

Also, I am not 100% sure about the overlay positioning, but I don't know where else to put it...

**Comment by P0keDev** [comment]
could you include a screenshot of the overlay in action?

**Comment by Incompleteusern** [comment]
image
from discord

**Comment by magicus** [comment]
> At the moment, if we disable this feature (and the debug one), do we get the old, unmodified, Wynncraft style NPC messages (with no modification, including the "shift to continue", and the grayed out stuff)?

That is the intention, yes. I have not verified it, but it it does not work like you describe, I'd consider it a bug and get it fixed. In fact, come to think of it, I'm not sure the logic is sound what happens if you enable the overlay but not the debug feature. I probably need to have like a "refcount" on how many features want to have NPC dialogue extractions enabled.

**Comment by magicus** [comment]
I'd prefer to fix that when I disable the debug feature by deafult, if that's okay.

**Comment by magicus** [comment]
As for positioning, I'd be fantastically cool if we could render the dialogues as speech bubbles floating in the world next to the characters. :-) At least, I think it would be cool. Maybe it'd just be impractical.

But I'm open for suggestion for other placements. To the left, just above the chat lines is another obvious choice. But I think that if you interact with an NPC, the dialog is kind of the most central thing to your play experience at that time, and the placement should mimic that.

**Comment by magicus** [comment]
I've now adapted the fix to the recently repaired calculateRenderHeight. I think it is about as clear for commit as it'll ever be.

File: common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java]: I think we can just not render a border for the overlay, it might look better. If you agree to do this, I will approve after this is changed, if you think we should keep this, LMK.

Comment (magicus) [reply common/src/main/java/com/wynntils/features/user/overlays/NpcDialogueOverlayFeature.java]: Let me check how that looks. I think it might be a bit bland. 

I'll be occupied elsewhere for a few days so it might be quiet for me for a while.

**Comment by kristofbolyai** [comment]
Also I would like to mention that this does not seem to work in production.

**Comment by magicus** [comment]
What is not working? 

And "in production", you mean as in contrast to "beta"? I have never run on beta, so I've tested it only on production. To be fair, I've only tested it with my level 11 test character, so just Detlas/Ragni NPCs, but they all worked fine.

**Comment by kristofbolyai** [comment]
> What is not working? 
> 
> 
> 
> And "in production", you mean as in contrast to "beta"? I have never run on beta, so I've tested it only on production. To be fair, I've only tested it with my level 11 test character, so just Detlas/Ragni NPCs, but they all worked fine.

Built jar in multimc, fabric loader + api. I assume it is because we don't have the debug feature.

**Comment by magicus** [comment]
Oh. Bummer. I guess I need to do that proper ref-count activation anyway, unless I can come up with some better solution.

**Comment by magicus** [comment]
Ok, i ditched the debug feature completely, and made sure the NPC dialogue extraction is enabled iff the NPC dialogue overlay is enabled.

I think this should be good to go in now.

**Comment by kristofbolyai** [comment]
Have you tested this works in production (some launcher)?

**Review by kristofbolyai — APPROVED** [review]
Seems fine

**Comment by magicus** [comment]
There should be no problem with production since I don't use debug features any longer. Do you want me to test it anyway?

**Comment by kristofbolyai** [comment]
Yes please just to make sure.

**Comment by magicus** [comment]
Working fine.

**Pull request merged**
